lab11.md

```diff
@@ -0,0 +1,184 @@
+# Lab 11 - Advanced feature - Trust with HMAC
+
+<img src="https://github.com/openfaas/media/raw/master/OpenFaaS_Magnet_3_1_png.png" width="500px"></img>
+
+Before starting this lab create new folder
+```
+mkdir -p lab11 \
+   && cd lab11
+```
+
+## What is HMAC
+
+Currently our functions can get information from everywhere without any restriction. If we want to validate the source of information we can use _Hash-based Message Authentication Code (HMAC)_.
+
+From [alexellis/hmac](https://github.com/alexellis/hmac):
+> HMAC uses a symmetric key that both sender/receiver share ahead of time. The sender will generate a hash when wanting to transmit a message - this data is sent along with the payload. The recipient will then sign payload with the shared key and if the hash matches then the payload is assumed to be from the sender.
+
+This way we close our functions from being invoked with invalid or even dangerous information.
+
+## Using HMAC
+
+We will use the `--sign` flag provided by faas-cli to send a header containing the message encrypted with the value of the `--key` flag.
+
+> Note: Both `--sign` and `--key` must be present.
+
+Let's first inspect what the flag does by deploying the `env` function which will print all of the environmental variables accessible inside the function:
+
+```
+$ faas-cli deploy --name env --fprocess="env" --image="functions/alpine:latest" --network=func_functions
+```
+
+* Invoke the function without `--sign` flag:
+
+```
+$ echo "The message" | faas-cli invoke env
+PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin
+HOSTNAME=d2c1a2cb20c2
+fprocess=env
+HOME=/root
+Http_X_Call_Id=b84947c6-2970-4fcf-ba3b-66dde6943999
+Http_X_Forwarded_For=10.255.0.2:34974
+Http_X_Forwarded_Host=127.0.0.1:8080
+Http_Content_Length=0
+Http_Accept_Encoding=gzip
+Http_Content_Type=text/plain
+Http_User_Agent=Go-http-client/1.1
+Http_X_Start_Time=1538725657952768349
+...
+```
+
+* Invoke the function again but this time with `--sign` flag:
+
+```
+$ echo -n "The message" | faas-cli invoke env --sign=hmac --key=cookie
+PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin
+HOSTNAME=d2c1a2cb20c2
+fprocess=env
+HOME=/root
+Http_User_Agent=Go-http-client/1.1
+Http_Content_Length=0
+Http_Accept_Encoding=gzip
+...
+Http_Hmac=sha1=9239edfe20185eafd7a5513c303b03d207d22f64
+...
+```
+
+We see the `hmac` value transformed to environmental variable `Http_Hmac` with encrypted request `The message` with key `cookie` as `9239edfe20185eafd7a5513c303b03d207d22f64` and encryption method `sha1` appended in front of the value.
+
+## HMAC in action
+
+For our purpose we are going to create a new Python 3 function. Let’s call it `hmac-protected`:
+
+```
+$ faas-cli new --lang python3 hmac-protected --prefix="<your-docker-username>"
+```
+
+Add `payload-secret` which will serve as the key that will encrypt the payload. 
+
+Populate `payload-secret` like we did in [lab10](https://github.com/openfaas/workshop/blob/master/lab10.md):
```

Comment:
Here we actually create it, so "populate" is not the best verb - it's used for something already created.